File: CSharpCodeBase/components/services/inputservice.cs
using System.Collections.Generic;
using InControl;
using UnityEngine;

namespace MainGame 
{
    public class InputService {
        private string defaultLayer = "Default Layer";
        private string mouseAndKeyboard = "Keyboard/Mouse";

        private Stack<string> inputLayersStack;
        private InputDevice inputDevice;
    /*
        public void init ()
        {
            inputLayersStack = new Stack<string>();
            inputLayersStack.Push(defaultLayer);
            inputDevice = InputManager.ActiveDevice;
        }
     */

        public bool IsGamepad()
        {
            return inputDevice.Name != mouseAndKeyboard;
        }

        public bool CheckLayer(string layer)
        {
            if (string.IsNullOrEmpty(layer))
            {
                layer = defaultLayer;
            }
            return layer == inputLayersStack.Peek();
        }

        public void PushFrame(string name)
        {
            inputLayersStack:Push(name);
        }


        public void PopFrame()
        {
            if (inputLayersStack.Peek() != defaultLayer)
            {
                inputLayersStack.Pop();
            }
        }

        public bool GetKey(KeyCode keycode, string layer=null)
        {
            if (CheckLayer(layer))
            {
                return Input.GetKey(keycode);
            }
            else
            {
                return false;
            }
        }


        public bool GetKeyDown(KeyCode keycode, string layer=null)
        {
            if(CheckLayer(layer))
            { 
                return Input.GetKeyDown(keycode);
            }
            else
            {
                return false;
            }
        }

        
        public bool GetKeyDownUnlocked(KeyCode keycode, string layer=null)
        {
            if (CheckLayer(layer))
            {
                return Input.GetKeyDown(keycode);
            }
            else
            {
                return false;
            }
        }

        public bool GetMouseButtonDown(int button, string layer=null)
        {
            if (CheckLayer(layer))
            {
                return Input.GetMouseButtonDown(button)
            }
            else
            {
                return false;
            }
        }

        public bool GetMouseButton(int button, string layer=null)
        {
            if (CheckLayer(layer))
            {
                return Input.GetMouseButtonDown(button);
            }
            else
            {
                return false;
            }
        }

        public bool BottomButtonWasPressed(string layer=null)
        {
            if (CheckLayer(layer))
            {
                return inputDevice.Action1.WasPressed;
            }
            else
            {
                return false;
            }
        }

        public bool LeftButtonWasPressed(string layer=null)
        {
            if (CheckLayer(layer))
            {
                return inputDevice.Action3.WasPressed;
            }
            else
            {
                return false;
            }
        }

        public bool TopButtonWasPressed(string layer=null)
        {
            if (CheckLayer(layer))
            {
                return inputDevice.Action4.WasPressed;
            }
            else
            {
                return false;
            }
        }

        public bool RightButtonWasPressed(string layer=null)
        {
            if (CheckLayer(layer))
            {
                return inputDevice.Action2.WasPressed;
            }
            else
            {
                return false;
            }
        }

        public Vector3 LeftStickValues(string layer=null)
        {
            Vector3 value = new Vector3();
            if (CheckLayer(layer))
            {
                value.x = inputDevice.LeftStickX.Value;
                value.z = inputDevice.LeftStickY.Value;
            }
            return value;
        }

        public bool LeftStickYIsPressed(string layer=null)
        {
            if (CheckLayer(layer))
            {
                return inputDevice.LeftStickY.IsPressed;
            }
            else
            {
                return false;
            }
        }

        public bool LeftStickXIsPressed(string layer=null)
        {
            if (CheckLayer(layer))
            {
                return inputDevice.LeftStickX.IsPressed;
            }
            else
            {
                return false;
            }
        }

        public bool RightTriggerWasPressed(string layer=null)
        {
            if (CheckLayer(layer))
            {
                return inputDevice.RightTrigger.WasPressed;
            }
            else
            {
                return false;
            }
        }

        public bool RightTriggerIsPressed(string layer=null)
        {
            if (CheckLayer(layer))
            {
                return inputDevice.RightTrigger.IsPressed;
            }
            else
            {
                return false;
            }
        }

        public bool LeftTriggerWasPressed(string layer=null)
        {
            if (CheckLayer(layer))
            {
                return inputDevice.LeftTrigger.WasPressed;
            }
            else
            {
                return false;
            }
        }

        public bool LeftTriggerIsPressed(string layer=null)
        {
            if (CheckLayer(layer))
            {
                return inputDevice.LeftTrigger.IsPressed;
            }
            else
            {
                return false;
            }
        }

        public Vector3 GetLookValue(string layer=null)
        {
            Vector3 value = new Vector3();
            if (CheckLayer(layer))
            {
                value.x = inputDevice.RightStickX.Value;
                value.z = inputDevice.RightStickY.Value;
            }
            return value;
        }

        // FIXME: Эти функции дублируются с двумя нижними
        public bool LookXWasPressed(string layer=null)
        {
            if (CheckLayer(layer))
            {
                return inputDevice.RightStickX.WasPressed;
            }
            else
            {
                return false;
            }
        }

        public bool LookYWasPressed(string layer=null)
        {
            if (CheckLayer(layer))
            {
                return inputDevice.RightStickY.WasPressed;
            }
            else
            {
                return false;
            }
        }

        public bool RightStickYWasPressed(string layer=null)
        {
            if (CheckLayer(layer))
            {
                return inputDevice.RightStickY.WasPressed;
            }
            else
            {
                return false;
            }
        }

        public bool RightStickXWasPressed(string layer=null)
        {
            if (CheckLayer(layer))
            {
                return inputDevice.RightStickX.WasPressed;
            }
            else
            {
                return false;
            }
        }

        public bool RightStickButtonWasPressed(string layer=null)
        {
            if (CheckLayer(layer))
            {
                return inputDevice.RightStickButton.WasPressed;
            }
            else
            {
                return false;
            }
        }

        public bool LeftStickButtonWasPressed(string layer=null)
        {
            if (CheckLayer(layer))
            {
                return inputDevice.LeftStickButton.WasPressed;
            }
            else
            {
                return false;
            }
        }

        public bool LeftBumperWasPressed(string layer=null)
        {
            if (CheckLayer(layer))
            {
                return inputDevice.LeftBumper.WasPressed;
            }
            else
            {
                return false;
            }
        }

        public bool LeftBumperIsPressed(string layer=null)
        {
            if (CheckLayer(layer))
            {
                return inputDevice.LeftBumper.IsPressed;
            }
            else
            {
                return false;
            }
        }

        public bool RightBumperWasPressed(string layer=null)
        {
            if (CheckLayer(layer))
            {
                return inputDevice.RightBumper.WasPressed;
            }
            else
            {
                return false;
            }
        }

        public bool RightBumperIsPressed(string layer=null)
        {
            if (CheckLayer(layer))
            {
                return inputDevice.RightBumper.IsPressed;
            }
            else
            {
                return false;
            }
        }

        public bool DPadLeftIsPressed(string layer=null)
        {
            if (CheckLayer(layer))
            {
                return inputDevice.DPadLeft.IsPressed;
            }
            else
            {
                return false;
            }
        }

        public bool DPadRightIsPressed(string layer=null)
        {
            if (CheckLayer(layer))
            {
                return inputDevice.DPadRight.IsPressed;
            }
            else
            {
                return false;
            }
        }

        public bool DPadUpIsPressed(string layer=null)
        {
            if (CheckLayer(layer))
            {
                return inputDevice.DPadUp.IsPressed;
            }
            else
            {
                return false;
            }
        }

        public bool DPadDawnIsPressed(string layer=null)
        {
            if (CheckLayer(layer))
            {
                return inputDevice.DPadDown.IsPressed;
            }
            else
            {
                return false;
            }
        }
    }
}

File: CSharpCodeBase/components/services/inputservice.lua
local Class = require 'utils.hump.class'
local Stack = require 'utils.hump.stack'
local Vector3 = require 'utils.hump.vector3'

local default = 'Default Layer'
local mouseAndKeyboard = 'Keyboard/Mouse'

local InputService = Class({init=function(self)
    self.inputLayersStack = Stack()
    self.inputLayersStack:Push(default)
    self.inputDevice = luanet.InControl.InputManager.ActiveDevice
end})

function InputService:IsGamepad()
  if self.inputDevice.Name == mouseAndKeyboard then
    return false
  end
  return true
end

function InputService:CheckLayer(layer)
  local mLayer = layer or default
  if mLayer and mLayer == self.inputLayersStack:Top() then
    return true
  end
  return false
end

function InputService:PushFrame(name)
  self.inputLayersStack:Push(name)
end

function InputService:PopFrame()
  if self.inputLayersStack:Top() ~= default then
    self.inputLayersStack:Pop()
  end
end

function InputService:GetKey(keycode, layer)
  if self:CheckLayer(layer) then
    return Input.GetKey(keycode)
  end
  return false
end

function InputService:GetKeyDown(keycode, layer)
  if self:CheckLayer(layer) then
    return Input.GetKeyDown(keycode)
  end
  return false
end

function InputService:GetKeyDownUnlocked(keycode, layer)
  if self:CheckLayer(layer) then
    return Input.GetKeyDownUnlocked(keycode)
  end
  return false
end

function InputService:GetMouseButtonDown(keycode, layer)
  if self:CheckLayer(layer) then
    return Input.GetMouseButtonDown(keycode)
  end
  return false
end

function InputService:GetMouseButton(keycode, layer)
  if self:CheckLayer(layer) then
    return Input.GetMouseButton(keycode)
  end
  return false
end

function InputService:BottomButtonWasPressed(layer)
  if self:CheckLayer(layer) then
    return self.inputDevice.Action1.WasPressed
  end
  return false
end

function InputService:LeftButtonWasPressed(layer)
  if self:CheckLayer(layer) then
    return self.inputDevice.Action3.WasPressed
  end
  return false
end

function InputService:TopButtonWasPressed(layer)
  if self:CheckLayer(layer) then
    return self.inputDevice.Action4.WasPressed
  end
  return false
end

function InputService:RightButtonWasPressed(layer)
  if self:CheckLayer(layer) then
    return self.inputDevice.Action2.WasPressed
  end
  return false
end

function InputService:LeftStickValues(layer)
  local value = Vector3()
  if self:CheckLayer(layer) then
    value.x = self.inputDevice.LeftStickX.Value
    value.z = self.inputDevice.LeftStickY.Value
  end
  return value
end

-- Левый стик был нажат по вертикале
function InputService:LeftStickYIsPressed(layer)
  if self:CheckLayer(layer) then
    return self.inputDevice.LeftStickY.IsPressed
  end
  return false
end

function InputService:LeftStickXIsPressed(layer)
  if self:CheckLayer(layer) then
    return self.inputDevice.LeftStickX.IsPressed
  end
  return false
end

function InputService:RightTriggerWasPressed(layer)
  if self:CheckLayer(layer) then
    return self.inputDevice.RightTrigger.WasPressed
  end
  return false
end

function InputService:RightTriggerIsPressed(layer)
  if self:CheckLayer(layer) then
    return self.inputDevice.RightTrigger.IsPressed
  end
  return false
end

function InputService:LeftTriggerWasPressed(layer)
  if self:CheckLayer(layer) then
    return self.inputDevice.LeftTrigger.WasPressed
  end
  return false
end

function InputService:LeftTriggerIsPressed(layer)
  if self:CheckLayer(layer) then
    return self.inputDevice.LeftTrigger.IsPressed
  end
  return false
end

function InputService:GetLookValue(layer)
  local value = Vector3()
  if self:CheckLayer(layer) then
    value.x = self.inputDevice.RightStickX.Value
    value.z = self.inputDevice.RightStickY.Value
  end
  return value
end

function InputService:LookXWasPressed(layer)
  if self:CheckLayer(layer) then
    return self.inputDevice.RightStickX.WasPressed
  end
  return false
end

function InputService:LookYWasPressed(layer)
  if self:CheckLayer(layer) then
    return self.inputDevice.RightStickY.WasPressed
  end
  return false
end

function InputService:RightStickYWasPressed(layer)
  if self:CheckLayer(layer) then
    return self.inputDevice.RightStickY.WasPressed
  end
  return false
end

function InputService:RightStickXWasPressed(layer)
  if self:CheckLayer(layer) then
    return self.inputDevice.RightStickX.WasPressed
  end
  return false
end

function InputService:RightStickButtonWasPressed(layer)
  if self:CheckLayer(layer) then
    return self.inputDevice.RightStickButton.WasPressed
  end
  return false
end

function InputService:LeftStickButtonWasPressed(layer)
  if self:CheckLayer(layer) then
    return self.inputDevice.LeftStickButton.WasPressed
  end
  return false
end

function InputService:LeftBumperWasPressed(layer)
  if self:CheckLayer(layer) then
    return self.inputDevice.LeftBumper.WasPressed
  end
  return false
end

function InputService:LeftBumperIsPressed(layer)
  if self:CheckLayer(layer) then
    return self.inputDevice.LeftBumper.IsPressed
  end
  return false
end

function InputService:RightBumperWasPressed(layer)
  if self:CheckLayer(layer) then
    return self.inputDevice.RightBumper.WasPressed
  end
  return false
end

function InputService:RightBumperIsPressed(layer)
  if self:CheckLayer(layer) then
    return self.inputDevice.RightBumper.IsPressed
  end
  return false
end

function InputService:DPadLeftIsPressed(layer)
  if self:CheckLayer(layer) then
    return self.inputDevice.DPadLeft.IsPressed
  end
  return false
end

function InputService:DPadRightIsPressed(layer)
  if self:CheckLayer(layer) then
    return self.inputDevice.DPadRight.IsPressed
  end
  return false
end

function InputService:DPadUpIsPressed(layer)
  if self:CheckLayer(layer) then
    return self.inputDevice.DPadUp.IsPressed
  end
  return false
end

function InputService:DPadDawnIsPressed(layer)
  if self:CheckLayer(layer) then
    return self.inputDevice.DPadDown.IsPressed
  end
  return false
end

return InputService